### PR TITLE
71 Global Exception Handler Middleware 

### PR DESCRIPTION
**Pull request by serhiiyasenev**
dev
## JIRA

* [Main JIRA ticket - 71](https://github.com/project-studying-dotnet/Streetcode-Server-June/issues/71)

### CODE REVIEW

File: Streetcode/Streetcode.BLL/Exceptions/RequestException.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Net;
+
+namespace Streetcode.BLL.Exceptions;
+
+public class RequestException : Exception
```
Comment (vladtymo) [inline Streetcode/Streetcode.BLL/Exceptions/RequestException.cs:5]: Add other common exception constructors
See: https://learn.microsoft.com/en-us/dotnet/standard/exceptions/how-to-create-user-defined-exceptions

File: Streetcode/Streetcode.BLL/MediatR/Media/Art/GetById/GetArtByIdHandler.cs [hunk 2]
```diff
@@ -28,9 +28,9 @@ public async Task<Result<ArtDTO>> Handle(GetArtByIdQuery request, CancellationTo
 
         if (art is null)
         {
-            string errorMsg = $"Cannot find an art with corresponding id: {request.Id}";
+            var errorMsg = $"Cannot find an art with corresponding id: {request.Id}";
             _logger.LogError(request, errorMsg);
-            return Result.Fail(new Error(errorMsg));
+            throw new RequestException(errorMsg, HttpStatusCode.NotFound);
```
Comment (vladtymo) [inline Streetcode/Streetcode.BLL/MediatR/Media/Art/GetById/GetArtByIdHandler.cs:33]: Remove these changes

Comment (vladtymo) [reply Streetcode/Streetcode.BLL/MediatR/Media/Art/GetById/GetArtByIdHandler.cs:33]: As we previously discussed, this task consider creating exception middleware only without changing the rising error method

File: Streetcode/Streetcode.XUnitTest/MediatRTests/Media/Art/GetArtByIdTest.cs
Comment (vladtymo) [inline Streetcode/Streetcode.XUnitTest/MediatRTests/Media/Art/GetArtByIdTest.cs:1]: Remove these changes

File: Streetcode/Streetcode.XUnitTest/MediatRTests/Partners/CreatePartnerHandlerTests.cs [hunk 3]
```diff
@@ -58,7 +58,7 @@ public CreatePartnerHandlerTests()
         }
 
         [Fact]
-        public async Task Handle_Should_ReturnEqualTrue_WhenInputNewPartner()
+        public void Handle_Should_ReturnEqualTrue_WhenInputNewPartner()
```
Comment (vladtymo) [inline Streetcode/Streetcode.XUnitTest/MediatRTests/Partners/CreatePartnerHandlerTests.cs:61]: Why these changes here?

File: Streetcode/Streetcode.WebApi/Middlewares/GenericExceptionHandlerMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    private static string SerializeJson<T>(T data)
+    {
+        var camelCasePropertiesSetting = new JsonSerializerSettings
+        {
+            ContractResolver = new CamelCasePropertyNamesContractResolver()
+        };
+
+        camelCasePropertiesSetting.Converters.Add(new Newtonsoft.Json.Converters.StringEnumConverter());
+
+        return JsonConvert.SerializeObject(data, camelCasePropertiesSetting);
```
Comment (vladtymo) [inline Streetcode/Streetcode.WebApi/Middlewares/GenericExceptionHandlerMiddleware.cs:50]: Why don't you use System.Text.Json built in .NET instead of Newtonsoft?

**Review by vladtymo — CHANGES_REQUESTED** [review]

File: Streetcode/Streetcode.BLL/MediatR/Media/Art/GetById/GetArtByIdHandler.cs (hunk 2 above)
Comment (serhiiyasenev) [reply Streetcode/Streetcode.BLL/MediatR/Media/Art/GetById/GetArtByIdHandler.cs:33]: done

File: Streetcode/Streetcode.WebApi/Middlewares/GenericExceptionHandlerMiddleware.cs (hunk 4 above)
Comment (serhiiyasenev) [reply Streetcode/Streetcode.WebApi/Middlewares/GenericExceptionHandlerMiddleware.cs:50]: because we have `Newtonsoft` installed - it usually means that we are going to use it instead of `System.Text.Json`

File: Streetcode/Streetcode.XUnitTest/MediatRTests/Media/Art/GetArtByIdTest.cs
Comment (serhiiyasenev) [reply Streetcode/Streetcode.XUnitTest/MediatRTests/Media/Art/GetArtByIdTest.cs:1]: done

File: Streetcode/Streetcode.XUnitTest/MediatRTests/Partners/CreatePartnerHandlerTests.cs (hunk 3 above)
Comment (serhiiyasenev) [reply Streetcode/Streetcode.XUnitTest/MediatRTests/Partners/CreatePartnerHandlerTests.cs:61]: to avoid warnings

File: Streetcode/Streetcode.BLL/Exceptions/RequestException.cs (hunk 1 above)
Comment (serhiiyasenev) [reply Streetcode/Streetcode.BLL/Exceptions/RequestException.cs:5]: added

Comment (vladtymo) [reply Streetcode/Streetcode.BLL/Exceptions/RequestException.cs:5]: With inner and status code and message also

**Review by vladtymo — APPROVED** [review]

File: Streetcode/Streetcode.BLL/Exceptions/RequestException.cs (hunk 1 above)
Comment (serhiiyasenev) [reply Streetcode/Streetcode.BLL/Exceptions/RequestException.cs:5]: done

File: Streetcode/Streetcode.BLL/Exceptions/RequestException.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public RequestException(HttpStatusCode statusCode)
```
Comment (BohdanLytvynov) [inline Streetcode/Streetcode.BLL/Exceptions/RequestException.cs:7]: Lines 7 - 20. It will be more useful to implement a chain of ctors using key word this. 
Here you can inspect good example https://stackoverflow.com/questions/1814953/how-to-do-constructor-chaining-in-c-sharp

**Review by BohdanLytvynov — CHANGES_REQUESTED** [review]
Modify code structure by creating a chain of ctors.

File: Streetcode/Streetcode.BLL/Exceptions/RequestException.cs (hunk 5 above)
Comment (serhiiyasenev) [reply Streetcode/Streetcode.BLL/Exceptions/RequestException.cs:7]: https://learn.microsoft.com/en-us/dotnet/standard/exceptions/how-to-create-user-defined-exceptions

I think they should be independent

**Review by BohdanLytvynov — APPROVED** [review]
